### PR TITLE
HPCC-31142 Encrypt agent->server packets in-place, and just once

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Using newly-improved udpsim program

### CODE REVIEW

**Review by ghalliday — APPROVED** [review]
Looks good to me.

**Comment by ghalliday** [comment]
@mckellyln please review - especially the second commit (I have merged a PR with the first).
@richardkchapman this will requite rebasing.

File: system/jlib/jencrypt.cpp [hunk 1]
```diff
@@ -1891,6 +1910,56 @@ MemoryBuffer &aesEncrypt(const void *key, size_t keylen, const void *plaintext,
...
 }
 
+extern jlib_decl size_t aesEncryptInPlace(const void *key, size_t keylen, void *buffer, size_t inlen, size_t buflen)
+{
+    if (!buffer || !inlen)
+        return 0;
+    byte *data = (byte *) buffer;
+    EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
+    if (!ctx)
+        encryptError("Failed to create context");
```
Comment (mckellyln) [inline system/jlib/jencrypt.cpp:1920]: if we throw here, is it ok to call EVP_CIPHER_CTX_free(ctx) without a valid ctx ?  Probably.

File: system/jlib/jencrypt.cpp [hunk 2]
```diff
@@ -1891,6 +1910,56 @@ MemoryBuffer &aesEncrypt(const void *key, size_t keylen, const void *plaintext,
...
+            case 16:
+                if(1 != EVP_EncryptInit_ex(ctx, EVP_aes_128_cbc(), NULL, (const unsigned char *) key, iv))
+                    encryptError("Failed to initialize context");
+                break;
+            default:
+                encryptError("Unsupported key length");
+                break;
+        }
+        // Make sure there is space for the padding. This is up to 16 bytes
+        if (buflen - inlen < 16)
```
Comment (mckellyln) [inline system/jlib/jencrypt.cpp:1943]: maybe put this check earlier, before bothering to init ?

**Review by mckellyln — APPROVED** [review]
First commit appears to have been merged already ?
Second commit looks ok - Approved - but I think needs rebasing since pull target branch was modified already for header length fix.

**Comment by richardkchapman** [comment]
@ghalliday Rebased and repushed

**Pull request merged**